### PR TITLE
cdc: capture old value from txn layer dynamically

### LINKED ISSUE #10091: CDC old value cache keeps grow after all regions are deregistered
## Bug Report

image
image

After all regions are deregistered, TiKV should add more old values to cache.

### Steps to reproduce

```
# create a changefeed with old value enabled.
sysbench --db-driver=mysql --mysql-host=172.16.5.33 --mysql-port=47904 --mysql-user=root --time=15000 --report-interval=10 --threads=64 --mysql-db=test --tables=1 --table-size=6000000 oltp_write_only prepare && \
sleep 60 && sysbench --db-driver=mysql --mysql-host=172.16.5.33 --mysql-port=47904 --mysql-user=root --time=15000 --report-interval=10 --threads=64 --mysql-db=test --tables=1 --table-size=6000000 oltp_write_only run
kill -SIGSTOP CDCPID
```

<details> <summary> TiKV log </summary>

```
[2021/04/28 10:02:01.058 +08:00] [INFO] [<unknown>] ["ipv4:172.16.5.33:45240: Keepalive watchdog fired. Closing transport."]
[2021/04/28 10:02:01.058 +08:00] [INFO] [service.rs:350] ["cdc receive closed"] [conn_id=ConnID(19)] [downstream=ipv4:172.16.5.33:45240]
[2021/04/28 10:02:01.058 +08:00] [WARN] [service.rs:375] ["cdc send failed"] [conn_id=ConnID(19)] [downstream=ipv4:172.16.5.33:45240] [error=RemoteStopped]
[2021/04/28 10:02:01.058 +08:00] [INFO] [endpoint.rs:344] ["cdc deregister"] [deregister="Deregister { deregister: \"conn\", conn_id: ConnID(19) }"]
[2021/04/28 10:02:01.058 +08:00] [INFO] [endpoint.rs:344] ["cdc deregister"] [deregister="Deregister { deregister: \"conn\", conn_id: ConnID(19) }"]
[2021/04/28 10:02:01.066 +08:00] [INFO] [<unknown>] ["ipv4:172.16.5.33:45242: Keepalive watchdog fired. Closing transport."]
[2021/04/28 10:02:01.066 +08:00] [INFO] [service.rs:350] ["cdc receive closed"] [conn_id=ConnID(20)] [downstream=ipv4:172.16.5.33:45242]
[2021/04/28 10:02:01.066 +08:00] [INFO] [endpoint.rs:344] ["cdc deregister"] [deregister="Deregister { deregister: \"conn\", conn_id: ConnID(20) }"]
[2021/04/28 10:02:01.066 +08:00] [WARN] [service.rs:375] ["cdc send failed"] [conn_id=ConnID(20)] [downstream=ipv4:172.16.5.33:45242] [error=RemoteStopped]
[2021/04/28 10:02:01.066 +08:00] [INFO] [endpoint.rs:344] ["cdc deregister"] [deregister="Deregister { deregister: \"conn\", conn_id: ConnID(20) }"]
[2021/04/28 10:02:03.538 +08:00] [INFO] [service.rs:350] ["cdc receive closed"] [conn_id=ConnID(21)] [downstream=ipv4:172.16.5.33:45262]
[2021/04/28 10:02:03.539 +08:00] [INFO] [endpoint.rs:344] ["cdc deregister"] [deregister="Deregister { deregister: \"conn\", conn_id: ConnID(21) }"]
[2021/04/28 10:02:03.539 +08:00] [WARN] [service.rs:375] ["cdc send failed"] [conn_id=ConnID(21)] [downstream=ipv4:172.16.5.33:45262] [error=RemoteStopped]
[2021/04/28 10:02:03.546 +08:00] [INFO] [endpoint.rs:344] ["cdc deregister"] [deregister="Deregister { deregister: \"conn\", conn_id: ConnID(21) }"]
[2021/04/28 10:02:03.870 +08:00] [INFO] [size.rs:168] ["approximate size over threshold, need to do split check"] [threshold=150994944] [size=156820425] [region_id=2]
[2021/04/28 10:02:04.361 +08:00] [INFO] [split_check.rs:302] ["update approximate size and keys with accurate value"] [keys=354265] [size=154807406] [region_id=2]
[2021/04/28 10:02:04.362 +08:00] [INFO] [peer.rs:3611] ["on split"] [source="split checker"] [split_keys="key 7480000000000000FF335F728000000005FF8FAA280000000000FAFA1BA7285603FFCA"] [peer_id=3] [region_id=2]
[2021/04/28 10:02:04.362 +08:00] [INFO] [pd.rs:554] ["try to batch split region"] [task=batch_split] [region="id: 2 start_key: 7480000000000000FF335F728000000005FF8DE8410000000000FA region_epoch { conf_ver: 1 version: 838 } peers { id: 3 store_id: 1 }"] [new_region_ids="[
[2021/04/28 10:02:04.363 +08:00] [INFO] [apply.rs:1292] ["execute admin command"] [command="cmd_type: BatchSplit splits { requests { split_key: 7480000000000000FF335F728000000005FF8FAA280000000000FA new_region_id: 1882 new_peer_ids: 1883 } right_derive: true }"] [index=80
[2021/04/28 10:02:04.363 +08:00] [INFO] [apply.rs:2131] ["split region"] [keys="key 7480000000000000FF335F728000000005FF8FAA280000000000FA"] [region="id: 2 start_key: 7480000000000000FF335F728000000005FF8DE8410000000000FA region_epoch { conf_ver: 1 version: 838 } peers {
[2021/04/28 10:02:04.364 +08:00] [INFO] [peer.rs:2223] ["notify pd with split"] [split_count=2] [peer_id=3] [region_id=2]
[2021/04/28 10:02:04.364 +08:00] [INFO] [peer.rs:2302] ["insert new region"] [region="id: 1882 start_key: 7480000000000000FF335F728000000005FF8DE8410000000000FA end_key: 7480000000000000FF335F728000000005FF8FAA280000000000FA region_epoch { conf_ver: 1 version: 839 } peers
[2021/04/28 10:02:04.364 +08:00] [INFO] [peer.rs:190] ["create peer"] [peer_id=1883] [region_id=1882]
[2021/04/28 10:02:04.364 +08:00] [INFO] [raft.rs:2443] ["switched to configuration"] [config="Configuration { voters: Configuration { incoming: Configuration { voters: {1883} }, outgoing: Configuration { voters: {} } }, learners: {}, learners_next: {}, auto_leave: false }
[2021/04/28 10:02:04.364 +08:00] [INFO] [raft.rs:1064] ["became follower at term 5"] [term=5] [raft_id=1883] [region_id=1882]
[2021/04/28 10:02:04.364 +08:00] [INFO] [raft.rs:375] [newRaft] [peers="Configuration { incoming: Configuration { voters: {1883} }, outgoing: Configuration { voters: {} } }"] ["last term"=5] ["last index"=5] [applied=5] [commit=5] [term=5] [raft_id=1883] [region_id=1882]
[2021/04/28 10:02:04.364 +08:00] [INFO] [raw_node.rs:285] ["RawNode created with id 1883."] [id=1883] [raft_id=1883] [region_id=1882]
[2021/04/28 10:02:04.364 +08:00] [INFO] [raft.rs:1488] ["starting a new election"] [term=5] [raft_id=1883] [region_id=1882]
[2021/04/28 10:02:04.364 +08:00] [INFO] [raft.rs:1114] ["became pre-candidate at term 5"] [term=5] [raft_id=1883] [region_id=1882]
[2021/04/28 10:02:04.364 +08:00] [INFO] [raft.rs:1088] ["became candidate at term 6"] [term=6] [raft_id=1883] [region_id=1882]
[2021/04/28 10:02:04.364 +08:00] [INFO] [raft.rs:1172] ["became leader at term 6"] [term=6] [raft_id=1883] [region_id=1882]
[2021/04/28 10:02:04.364 +08:00] [INFO] [peer.rs:3611] ["require updating max ts"] [initial_status=25769805454] [region_id=1882]
[2021/04/28 10:02:04.364 +08:00] [INFO] [endpoint.rs:223] ["register observe region"] [region="id: 1882 start_key: 7480000000000000FF335F728000000005FF8DE8410000000000FA end_key: 7480000000000000FF335F728000000005FF8FAA280000000000FA region_epoch { conf_ver: 1 version: 83
[2021/04/28 10:02:04.364 +08:00] [INFO] [pd.rs:982] ["succeed to update max timestamp"] [region_id=1882]
[2021/04/28 10:02:04.365 +08:00] [INFO] [scheduler.rs:391] ["get snapshot failed"] [err="Request(message: \"EpochNotMatch current epoch of region 2 is conf_ver: 1 version: 839, but you sent conf_ver: 1 version: 838\" epoch_not_match { current_regions { id: 2 start_key: 74
[2021/04/28 10:02:04.365 +08:00] [INFO] [scheduler.rs:391] ["get snapshot failed"] [err="Request(message: \"EpochNotMatch current epoch of region 2 is conf_ver: 1 version: 839, but you sent conf_ver: 1 version: 838\" epoch_not_match { current_regions { id: 2 start_key: 74
[2021/04/28 10:02:04.365 +08:00] [INFO] [scheduler.rs:391] ["get snapshot failed"] [err="Request(message: \"EpochNotMatch current epoch of region 2 is conf_ver: 1 version: 839, but you sent conf_ver: 1 version: 838\" epoch_not_match { current_regions { id: 2 start_key: 74
[2021/04/28 10:02:04.365 +08:00] [INFO] [scheduler.rs:391] ["get snapshot failed"] [err="Request(message: \"EpochNotMatch current epoch of region 2 is conf_ver: 1 version: 839, but you sent conf_ver: 1 version: 838\" epoch_not_match { current_regions { id: 2 start_key: 74
[2021/04/28 10:02:04.365 +08:00] [INFO] [scheduler.rs:391] ["get snapshot failed"] [err="Request(message: \"EpochNotMatch current epoch of region 2 is conf_ver: 1 version: 839, but you sent conf_ver: 1 version: 838\" epoch_not_match { current_regions { id: 2 start_key: 74
[2021/04/28 10:02:10.562 +08:00] [INFO] [<unknown>] ["ipv4:172.16.5.33:45203: Keepalive watchdog fired. Closing transport."]
[2021/04/28 10:02:10.562 +08:00] [INFO] [service.rs:350] ["cdc receive closed"] [conn_id=ConnID(18)] [downstream=ipv4:172.16.5.33:45203]
[2021/04/28 10:02:10.562 +08:00] [INFO] [endpoint.rs:344] ["cdc deregister"] [deregister="Deregister { deregister: \"conn\", conn_id: ConnID(18) }"]
[2021/04/28 10:02:10.562 +08:00] [WARN] [service.rs:375] ["cdc send failed"] [conn_id=ConnID(18)] [downstream=ipv4:172.16.5.33:45203] [error=RemoteStopped]
[2021/04/28 10:02:10.562 +08:00] [INFO] [endpoint.rs:344] ["cdc deregister"] [deregister="Deregister { deregister: \"conn\", conn_id: ConnID(18) }"]
[2021/04/28 10:02:10.562 +08:00] [INFO] [<unknown>] ["ipv4:172.16.5.33:45206: Keepalive watchdog fired. Closing transport."]
[2021/04/28 10:02:10.562 +08:00] [INFO] [service.rs:350] ["cdc receive closed"] [conn_id=ConnID(17)] [downstream=ipv4:172.16.5.33:45206]
[2021/04/28 10:02:10.562 +08:00] [WARN] [service.rs:375] ["cdc send failed"] [conn_id=ConnID(17)] [downstream=ipv4:172.16.5.33:45206] [error=RemoteStopped]
[2021/04/28 10:02:10.562 +08:00] [INFO] [endpoint.rs:344] ["cdc deregister"] [deregister="Deregister { deregister: \"conn\", conn_id: ConnID(17) }"]
[2021/04/28 10:02:10.562 +08:00] [INFO] [endpoint.rs:344] ["cdc deregister"] [deregister="Deregister { deregister: \"conn\", conn_id: ConnID(17) }"]
[2021/04/28 10:02:11.196 +08:00] [INFO] [tracker.rs:248] [slow-query] [perf_stats.internal_delete_skipped_count=0] [perf_stats.internal_key_skipped_count=230567] [perf_stats.block_read_byte=58208852] [perf_stats.block_read_count=1510] [perf_stats.block_cache_hit_count=28]
[2021/04/28 10:02:28.106 +08:00] [INFO] [tracker.rs:248] [slow-query] [perf_stats.internal_delete_skipped_count=0] [perf_stats.internal_key_skipped_count=230349] [perf_stats.block_read_byte=58226987] [perf_stats.block_read_count=1514] [perf_stats.block_cache_hit_count=26]
[2021/04/28 10:02:43.875 +08:00] [INFO] [size.rs:168] ["approximate size over threshold, need to do split check"] [threshold=150994944] [size=162088236] [region_id=2]
[2021/04/28 10:02:44.582 +08:00] [INFO] [split_check.rs:302] ["update approximate size and keys with accurate value"] [keys=374075] [size=163454334] [region_id=2]
[2021/04/28 10:02:44.583 +08:00] [INFO] [peer.rs:3611] ["on split"] [source="split checker"] [split_keys="key 7480000000000000FF335F728000000005FF916C0F0000000000FAFA1BA725FE03FFE8"] [peer_id=3] [region_id=2]
[2021/04/28 10:02:44.583 +08:00] [INFO] [pd.rs:554] ["try to batch split region"] [task=batch_split] [region="id: 2 start_key: 7480000000000000FF335F728000000005FF8FAA280000000000FA region_epoch { conf_ver: 1 version: 839 } peers { id: 3 store_id: 1 }"] [new_region_ids="[
[2021/04/28 10:02:44.584 +08:00] [INFO] [apply.rs:1292] ["execute admin command"] [command="cmd_type: BatchSplit splits { requests { split_key: 7480000000000000FF335F728000000005FF916C0F0000000000FA new_region_id: 1884 new_peer_ids: 1885 } right_derive: true }"] [index=80
[2021/04/28 10:02:44.584 +08:00] [INFO] [apply.rs:2131] ["split region"] [keys="key 7480000000000000FF335F728000000005FF916C0F0000000000FA"] [region="id: 2 start_key: 7480000000000000FF335F728000000005FF8FAA280000000000FA region_epoch { conf_ver: 1 version: 839 } peers {
[2021/04/28 10:02:44.584 +08:00] [INFO] [peer.rs:2223] ["notify pd with split"] [split_count=2] [peer_id=3] [region_id=2]
[2021/04/28 10:02:44.584 +08:00] [INFO] [peer.rs:2302] ["insert new region"] [region="id: 1884 start_key: 7480000000000000FF335F728000000005FF8FAA280000000000FA end_key: 7480000000000000FF335F728000000005FF916C0F0000000000FA region_epoch { conf_ver: 1 version: 840 } peers
```

</details>

### What did you expect?

No more old values are added to the cache after regions deregistered.

### PR DESCRIPTION
**Pull request by hicqu**
### What is changed and how it works?

Close #10091

* remove stuff about the `enable_old_value` option, because it will be always `true` now;
* refactor `Delegate` to make it easiler to maintain;
* avoid to capture old vaue from txn layer if all downstreams are stopped.

### Related changes

- PR to update `pingcap/docs`/`pingcap/docs-cn`:
- PR to update `pingcap/tidb-ansible`:
- Need to cherry-pick to the release branch

### Check List 

Tests 

- Integration test

Side effects

### Release note 

```release-note
None
```

### CODE REVIEW

**Review by overvenus — COMMENTED** [review]
Could you summarize changes in PR description? It would be helpful for reviewers.

File: components/cdc/src/delegate.rs [hunk 1]
```diff
@@ -192,108 +182,75 @@ pub struct Delegate {
     region: Option<Region>,
     pub resolver: Option<Resolver>,
 
-    pub downstreams: Vec<Downstream>,
+    downstreams: Vec<Downstream>,
     pending: Option<Pending>,
+    txn_extra_op: Arc<AtomicCell<TxnExtraOp>>,
     failed: bool,
-    pub txn_extra_op: TxnExtraOp,
+    old_value_requirers: usize,
```
[the remark this answers is not in the document]
Comment (hicqu) [reply components/cdc/src/delegate.rs]: Done. PTAL thanks!

File: components/cdc/src/delegate.rs [hunk 2]
```diff
@@ -309,9 +261,10 @@ impl Delegate {
     /// This means the region has met an unrecoverable error for CDC.
     /// It broadcasts errors to all downstream and stops.
     pub fn stop(&mut self, err: Error) {
+        self.handle.stop_observing();
```
Comment (overvenus) [inline components/cdc/src/delegate.rs:264]:
```suggestion
        // Stop observe further events.
        self.handle.stop_observing();
```

File: components/cdc/src/delegate.rs [hunk 3]
```diff
@@ -309,9 +261,10 @@ impl Delegate {
...
+        self.txn_extra_op.store(TxnExtraOp::Noop);
```
Comment (overvenus) [inline components/cdc/src/delegate.rs:265]: Please also add some comments about this line.

File: components/cdc/src/delegate.rs [hunk 4]
```diff
@@ -741,6 +702,48 @@ impl Delegate {
...
+        self.txn_extra_op.store(TxnExtraOp::ReadOldValue);
+    }
+
+    fn remove_downstream(&mut self, id: DownstreamID) -> Option<Downstream> {
+        let downstreams = self.downstreams_mut();
+        if let Some(index) = downstreams.iter().position(|x| x.id == id) {
+            let downstream = downstreams.swap_remove(index);
+            if self.downstreams.is_empty() {
+                self.handle.stop_observing();
+                self.txn_extra_op.store(TxnExtraOp::Noop);
```
Comment (overvenus) [inline components/cdc/src/delegate.rs:717]: It's similar to `stop`, can we unify them?

File: components/cdc/src/endpoint.rs [hunk 5]
```diff
@@ -476,7 +470,8 @@ impl<T: 'static + RaftStoreRouter<E>, E: KvEngine> Endpoint<T, E> {
                     conn.take_downstreams().into_iter().for_each(
                         |(region_id, (downstream_id, _))| {
                             if let Some(delegate) = self.capture_regions.get_mut(&region_id) {
-                                if delegate.unsubscribe(downstream_id, None) {
+                                delegate.unsubscribe(downstream_id, None);
```
Comment (overvenus) [inline components/cdc/src/endpoint.rs:473]: Looks like this PR fixed #10091 where when a TCP is disconnected, txn layer still adds old value to cache. 

Do we have tests to cover this case?

File: components/cdc/src/endpoint.rs [hunk 6]
```diff
@@ -1110,7 +1112,7 @@ impl<E: KvEngine> Initializer<E> {
             .fill_cache(false)
             .range(None, None)
             .hint_min_ts(hint_min_ts)
-            .build_delta_scanner(self.checkpoint_ts, self.txn_extra_op)
+            .build_delta_scanner(self.checkpoint_ts, TxnExtraOp::ReadOldValue)
```
[the remark this answers is not in the document]
Comment (hicqu) [reply components/cdc/src/endpoint.rs:1115]: I prefer to keep `build_delta_scanner` because it's in transaction module, maybe other places will need different old value options later.

File: components/cdc/src/delegate.rs [hunk 7]
```diff
@@ -192,108 +179,73 @@ pub struct Delegate {
     region: Option<Region>,
     pub resolver: Option<Resolver>,
 
-    pub downstreams: Vec<Downstream>,
+    downstreams: Vec<Downstream>,
     pending: Option<Pending>,
+    txn_extra_op: Arc<AtomicCell<TxnExtraOp>>,
```
[the remark this answers is not in the document]
Comment (hicqu) [reply components/cdc/src/delegate.rs:190]: It's commentted on `fn txn_extra_op` now.

File: components/cdc/src/endpoint.rs [hunk 8]
```diff
@@ -506,6 +501,15 @@ impl<T: 'static + RaftStoreRouter<E>, E: KvEngine> Endpoint<T, E> {
         let region_id = request.region_id;
         let downstream_id = downstream.get_id();
         let downstream_state = downstream.get_state();
+
+        let txn_extra_op = match self.get_txn_extra_op(region_id) {
+            Some(txn_extra_op) => txn_extra_op,
+            None => {
+                error!("cdc register for a not found region"; "region_id" => region_id);
+                return;
```
[the remark this answers is not in the document]
Comment (hicqu) [reply components/cdc/src/endpoint.rs:514]: Done.

File: components/cdc/src/endpoint.rs (hunk 5 above)
Comment (hicqu) [reply components/cdc/src/endpoint.rs:473]: Done.

File: components/cdc/src/endpoint.rs (hunk 8 above)
[the remark this answers is not in the document]
Comment (overvenus) [reply components/cdc/src/endpoint.rs:514]: Could you check cluster id mismatch first?

**Review by overvenus — COMMENTED** [review]
Rest LGTM

**Review by overvenus — APPROVED** [review]

**Review by hi-rustin — APPROVED** [review]
Looks good.

File: components/cdc/src/delegate.rs [hunk 9]
```diff
@@ -741,6 +711,54 @@ impl Delegate {
...
+    fn add_downstream(&mut self, downstream: Downstream) {
+        self.downstreams_mut().push(downstream);
+        self.txn_extra_op.store(TxnExtraOp::ReadOldValue);
+    }
+
+    fn remove_downstream(&mut self, id: DownstreamID) -> Option<Downstream> {
+        let downstreams = self.downstreams_mut();
+        if let Some(index) = downstreams.iter().position(|x| x.id == id) {
+            let downstream = downstreams.swap_remove(index);
+            if self.downstreams.is_empty() {
```
Comment (hicqu) [inline components/cdc/src/delegate.rs:724]: It's a bug. Should be `downstreams` instead of `self.downstreams`.
